### ISSUE TITLE
Compile error with MongoDB & required relationships

### ISSUE THREAD

**Issue by hdurix** [jhipster/generator-jhipster#9820]
##### **Overview of the issue**
When using MongoDB as database and entities with required relationships between them, some Integration Tests don't compile.

##### **Motivation for or Use Case**

It's clearly a bug as it does not compile :)

##### **Reproduce the error**
Generate an app with MongoDB and required relationships (see JDL below)

##### **Related issues**
Issue: https://github.com/jhipster/generator-jhipster/issues/9666
PR: https://github.com/jhipster/generator-jhipster/pull/9669

ping @Gabrui 

##### **Suggest a Fix**
The problem is that we use `TestUtil.findAll(em, ...)` [here](https://github.com/jhipster/generator-jhipster/blob/master/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs#L457) in every case (MongoDB included) but this method is only defined when using `sql` database: see [this condition](https://github.com/jhipster/generator-jhipster/blob/master/generators/server/templates/src/test/java/package/web/rest/TestUtil.java.ejs#L176).

I tried to fix it but it's a bit tricky, what I had in mind:
- use other relationship entity repository instead of the newly created `TestUtil.findAll(em, ...)` method

I only mentioned MongoDB here, but I'm sure it's the same for `Couchbase` and `Cassandra`.

> NB: maybe this configuration is missing in CI tests: MongoDB + required relationship.

##### **JHipster Version(s)**

JHipster from master (6.0.1)
```
4@0.0.0 /home/hippolyte/workspace/jhipster/samples/mongoRequired
└── (empty)

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.mycompany.myapp",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "6.0.1",
    "applicationType": "monolith",
    "baseName": "4",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "no",
    "enableHibernateCache": false,
    "websocket": false,
    "databaseType": "mongodb",
    "devDatabaseType": "mongodb",
    "prodDatabaseType": "mongodb",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "bXktc2VjcmV0LXRva2VuLXRvLWNoYW5nZS1pbi1wcm9kdWN0aW9uLWFuZC10by1rZWVwLWluLWEtc2VjdXJlLXBsYWNl",
    "useSass": true,
    "clientPackageManager": "npm",
    "clientFramework": "angularX",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "otherModules": [],
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": ["en"]
  }
}

</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Store {
  name String
}
entity Rule
relationship OneToMany {
  Store{rule} to Rule{store(name) required}
}

</pre>
</details>

##### **Environment and Tools**

openjdk version "1.8.0_212"
OpenJDK Runtime Environment (build 1.8.0_212-8u212-b03-0ubuntu1.18.04.1-b03)
OpenJDK 64-Bit Server VM (build 25.212-b03, mixed mode)

git version 2.17.1

node: v10.15.3

npm: 6.4.1

yeoman: 2.0.2

yarn: 1.6.0

Docker version 18.06.3-ce, build d7080c1

docker-compose version 1.21.1, build 5a3f1a3

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>Entity Rule definition</summary>

<pre>
{
    "name": "Rule",
    "fields": [],
    "relationships": [
        {
            "relationshipType": "many-to-one",
            "otherEntityName": "store",
            "otherEntityRelationshipName": "rule",
            "relationshipValidateRules": "required",
            "relationshipName": "store",
            "otherEntityField": "name"
        }
    ],
    "changelogDate": "20190529133335",
    "entityTableName": "rule",
    "dto": "no",
    "pagination": "no",
    "service": "no",
    "jpaMetamodelFiltering": false,
    "fluentMethods": true,
    "clientRootFolder": "",
    "applications": "*"
}
</pre>
</details>

<details>
<summary>Entity Store definition</summary>

<pre>
{
    "name": "Store",
    "fields": [
        {
            "fieldName": "name",
            "fieldType": "String"
        }
    ],
    "relationships": [
        {
            "relationshipType": "one-to-many",
            "otherEntityName": "rule",
            "otherEntityRelationshipName": "store",
            "relationshipName": "rule"
        }
    ],
    "changelogDate": "20190529133334",
    "entityTableName": "store",
    "dto": "no",
    "pagination": "no",
    "service": "no",
    "jpaMetamodelFiltering": false,
    "fluentMethods": true,
    "clientRootFolder": "",
    "applications": "*"
}
</pre>
</details>

##### **Browsers and Operating System**

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by Gabrui** [comment]
Are unique constraints enforced in mongodb? If not, we could just force my fix to the sql case. Just changing [this line](https://github.com/jhipster/generator-jhipster/blob/507e89c98f9c17e9787066c5229011e1d74ff709/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs#L446)
```
<%_ if (otherEntityName === 'user') { // TODO or other entity has no unique fields _%>
``` 

to

```
<%_ if (otherEntityName === 'user' || databaseType === 'sql') { // TODO or other entity has no unique fields _%>
``` 

would resolve.

But I really don't know how those other database types work.

**Comment by hdurix** [comment]
It's compiling but some tests are failing, I guess that's because we don't create the associated relationships entities during the tests, but I don't know how MongoDB relationships are working either.

I used the same JDL as you @Gabrui:
<details>
<summary>JDL with required definitions</summary>

<pre>
entity BankAccount {
  name String required
  bankNumber Integer
  agencyNumber Long
  lastOperationDuration Float
  meanOperationDuration Double
  balance BigDecimal required
  openingDay LocalDate
  lastOperationDate Instant
  active Boolean
  accountType BankAccountType
  attachment AnyBlob
  description TextBlob
}
entity TheLabel {
  labelName String required minlength(3)
}
entity Operation {
  date Instant required
  description String
  amount BigDecimal required
}

enum BankAccountType {
  CHECKING,
  SAVINGS,
  LOAN
}

entity Department {
  name String required,
  description TextBlob,
  advertisement Blob,
  logo ImageBlob
}

/**
 * JobHistory comment.
 */
entity JobHistory {
  startDate ZonedDateTime,
  endDate ZonedDateTime,
  language Language
}

enum Language {
  FRENCH, ENGLISH, SPANISH
}

enum JobType {
  BOSS, SLAVE
}

entity Job {
  title String minlength(5) maxlength(25),
  type JobType,
  minSalary Long,
  maxSalary Long
}

/**
 * The Employee entity.
 * Second line in javadoc.
 */
entity Employee {
  /**
   * The firstname attribute.
   */
  firstName String,
  lastName String,
  email String,
  phoneNumber String,
  hireDate ZonedDateTime,
  salary Long,
  commissionPct Long
}

entity Location {
  streetAddress String,
  postalCode String,
  city String,
  stateProvince String
}

entity Task {
  title String,
  description String
}

entity GoldenBadge {
  name String
}

entity SilverBadge {
  name String
}

entity Identifier {
  name String required unique
}

entity Country {
  name String
}

entity Region {
  name String
}

relationship OneToOne {
  Department{location} to Location,
  Employee{user(login)} to User with jpaDerivedIdentifier
}

relationship OneToMany {
  BankAccount{operation} to Operation{bankAccount(name)}
}
relationship ManyToOne {
  BankAccount{user(login)} to User
}
relationship ManyToMany {
  Operation{theLabel(labelName)} to TheLabel{operation}
}

relationship OneToMany {
  /**
   * A relationship
   */
  Department{employee} to
  /**
   * Another side of the same relationship,
   */
  Employee{department},
  Employee{job} to Job{emp(lastName)},
  Location{country} to Country,
  Country{area(name)} to Region
}

relationship ManyToOne {
  Employee{manager(lastName)} to Employee,
  Employee{sibag(name) required} to SilverBadge,
  Employee{gobag(name) required} to GoldenBadge,
  SilverBadge{iden(name) required} to Identifier,
  GoldenBadge{iden(name) required} to Identifier
}

relationship ManyToMany {
  JobHistory to Department,
  JobHistory to Job{history},
  JobHistory{emp(firstName)} to Employee{history},
  Job{chore(title)} to Task{linkedJob(title)}
}


angularSuffix BankAccount with mySuffix
filter BankAccount, Employee
clientRootFolder BankAccount, TheLabel, Operation with test-root

paginate TheLabel, Job with pagination
paginate Operation, JobHistory, Employee with infinite-scroll

service TheLabel, Employee, Department, Region with serviceClass
service BankAccount, Location, Country with serviceImpl

</pre>
</details>

@ivangsa if you have got an idea.

**Comment by Gabrui** [comment]
Then we need a way to make a query in those other databases. I remember that the first time I tried to fix that bug I imported the entitie's Repository but I also ran into the problem of the persistence context (it needed the @Transactional decorator to work) but it also gave other problems so I used the 'em' to make the query and simplify.

**Comment by ivangsa** [comment]
Hi, I will look into it..
There is also a compilation problem in services related to the primary key not being an string...
https://github.com/jhipster/generator-jhipster/pull/9826

**Comment by hdurix** [comment]
Yes @Gabrui the other issue using repositories inside the `createEntity` method is that this method is static (in order to be called from other integration tests) so the `@Autowired` repositories can't be used here.

@ivangsa thanks! I also noticed this compilation problem with ids using services (with the last JDL I gave).

**Comment by ivangsa** [comment]
after aplying fix proposed by @Gabrui ,
I think the remaining failing tests are because code is generated for "isUsingMapsId === true" which does not apply to mondgodb... harcoding that to false makes generated tests to pass...

**Comment by ivangsa** [comment]
hi, I think failing tests are caused by this (incompatible) setting in the jdl:
_Employee{user(login)} to User with jpaDerivedIdentifier_ 
this is generating code in services and tests for MapIds incompatible with mongodb..

**Comment by nonomoho** [comment]
For me there are some changes to do in `EntityResourceIT.java.ejs` file:
[This line](https://github.com/jhipster/generator-jhipster/blob/507e89c98f9c17e9787066c5229011e1d74ff709/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs#L446):
`if (TestUtil.findAll(em, <%= otherEntityNameCapitalized %>.class).isEmpty()) {`
And the "else" part:
```java
} else {
    <%= otherEntityName %> = TestUtil.findAll(em, <%= otherEntityNameCapitalized %>.class).get(0);
}
```
 should only be used for the SQL databaseType.

I can try to push something!

**Comment by hdurix** [comment]
@ivangsa sorry, I just copied the JDL and changed the databaseType without checking it was correct.

Both fixes are correct, the one from @nonomoho can be easier to read later.

**Comment by hdurix** [comment]
Fixed by #9830.
